core/lib/src/fs/static_file.rs

```diff
@@ -0,0 +1,495 @@
+
+            self.apply_response_headers(&mut builder, Status::PartialContent);
+
+            if from > 0 {
+                std::pin::Pin::new(&mut self.file)
+                    .start_seek(SeekFrom::Start(from))
+                    .unwrap();
+            }
+
+            builder.sized_body(length as usize, self.file);
+
+            Ok(builder.finalize())
+        } else {
+            // 14.16 Content-Range - A server sending a response with status code 416 (Requested range not satisfiable)
+            // SHOULD include a Content-Range field with a byte-range-resp-spec of "*". The instance-length specifies
+            // the current length of the selected resource.  e.g. */length
+            let mut builder = Response::build();
+            builder.header(ContentRangeHeaderValue::from(self.len));
+
+            self.apply_response_headers(&mut builder,Status::RangeNotSatisfiable);
+
+            Ok(builder.finalize())
+        }
+    }
+}
+
+/// Streams the named file to the client. Sets or overrides the Content-Type in
+/// the response according to the file's extension if the extension is
+/// recognized. See [`ContentType::from_extension()`] for more information. If
+/// you would like to stream a file with a different Content-Type than that
+/// implied by its extension, use a [`File`] directly.
+impl<'r> Responder<'r, 'static> for StaticFile {
+    fn respond_to(mut self, req: &'r Request<'_>) -> response::Result<'static> {
+        use PreconditionState::*;
+        self.comprehend_request_headers(req);
+        match self.get_precondition_state() {
+            Unspecified | ShouldProcess=> {
+                if req.method() == Method::Head {
+                    Ok(Response::build()
+                        .status(Status::Ok)
+                        .finalize())
+                } else{
+                    if self.is_range_request() {
+                        self.send_range(req)
+                    } else {
+                        self.send(req)
+                    }
+                }
+            }
+            NotModified => {
+                Ok(Response::build()
+                        .status(Status::NotModified)
+                        .finalize())
+            }
+            PreconditionFailed => {
+                Ok(Response::build()
+                    .status(Status::PreconditionFailed)
+                    .finalize())
+            }
+        }
+    }
+}
+
+impl Deref for StaticFile {
+    type Target = File;
+
+    fn deref(&self) -> &File {
+        &self.file
+    }
+}
+
+impl DerefMut for StaticFile {
+    fn deref_mut(&mut self) -> &mut File {
+        &mut self.file
+    }
+}
+
+
+#[repr(u8)]
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Ord, PartialOrd)]
+enum PreconditionState
+{
+    Unspecified,
+    NotModified,
+    ShouldProcess,
+    PreconditionFailed,
+}
+
+#[repr(u8)]
+#[derive(EnumFlags, Copy, Clone, Eq, PartialEq)]
+enum RequestType {
```

Comment:
Can you get rid of `EnumFlags` ? I got undefined behavior  in release mode (in debug mode everything works fine).

But on release i got strange behavior e.g.
```
---request begin---
GET /software/myfile HTTP/1.1
User-Agent: Wget/1.20.3 (linux-gnu)
Accept: */*
Accept-Encoding: identity
Host: 127.0.0.1:9854
Connection: Keep-Alive

---request end---
Żądanie HTTP wysłano, oczekiwanie na odpowiedź... 
---response begin---
HTTP/1.1 416 Range Not Satisfiable
content-range: bytes */79355392
server: Rocket
x-frame-options: SAMEORIGIN
x-content-type-options: nosniff
permissions-policy: interest-cohort=()
content-length: 0
date: Wed, 04 Aug 2021 14:21:10 GMT
```

There should not be `Range Not Satisfiable` in this situation (It's 200 OK in debug mode).


After ding a little I found problem in lines:
```rust
if is_range_request {
    self.request_type |= RequestType::IsRange
} else {
    self.request_type -= RequestType::IsRange
}
```
When `is_range_request == false` it entered in first branch and `self.request_type` was set to `RequestType::IsRange`. After adding some `pirntln!()`` I got core dump.

I rewrote `RequestType` to 

```rust
#[derive(Copy, Debug, Clone, Eq, PartialEq, Default)]
struct RequestType {
    _is_range: bool,
}

impl RequestType {
    fn is_range(&self) -> bool {
        self._is_range
    }

    fn set_range(&mut self, v: bool) {
        self._is_range = v;
    }
}
```
and it's started to work in release and debug mode.

In fact you don't need enum `RequestType` at all. Fields other that `IsRange` are not used.

Reply:
Could you send your code above as PR to my  branch? I'am busy these days.